swarm/chunk/tags.go

```diff
@@ -0,0 +1,233 @@
+func (ts *tags) Inc(s string, f State) {
+	t, ok := ts.tags.Load(s)
+	if !ok {
+		return
+	}
+	t.(*Tag).Inc(f)
+}
+
+// Get returns the state count for a tag
+func (ts *tags) Get(s string, f State) int {
+	t, _ := ts.tags.Load(s)
+	return t.(*Tag).Get(f)
+}
+
+// State is the enum type for chunk states
+type State = uint32
+
+const (
+	SPLIT  State = iota // chunk has been processed by filehasher/swarm safe call
+	STORED              // chunk stored locally
+	SENT                // chunk sent to neighbourhood
+	SYNCED              // proof is received; chunk removed from sync db; chunk is available everywhere
+)
+
+// Tag represents info on the status of new chunks
+type Tag struct {
+	uid       uint32     // a unique identifier for this tag
+	name      string     // a name tag for this tag
+	total     uint32     // total chunks belonging to a tag
+	split     uint32     // number of chunks already processed by splitter for hashing
+	stored    uint32     // number of chunks already stored locally
+	sent      uint32     // number of chunks sent for push syncing
+	synced    uint32     // number of chunks synced with proof
+	startedAt time.Time  // tag started to calculate ETA
+	State     chan State // channel to signal completion
+}
+
+// New creates a new tag, stores it by the name and returns it
+// it returns an error if the tag with this name already exists
+func NewTag(uid uint32, s string, total uint32) *Tag {
+	if len(s) == 0 {
+		s = fmt.Sprintf("upload_%d", time.Now().UnixNano())
+	}
+	t := &Tag{
+		uid:       uid,
+		name:      s,
+		startedAt: time.Now(),
+		total:     uint32(total),
+		State:     make(chan State, 5),
+	}
+	return t
+}
+
+// Inc increments the count for a state
+func (t *Tag) Inc(state State) {
+	var v *uint32
+	switch state {
+	case SPLIT:
+		v = &t.split
+	case STORED:
+		v = &t.stored
+	case SENT:
+		v = &t.sent
+	case SYNCED:
+		v = &t.synced
+	}
+	n := atomic.AddUint32(v, 1)
+	if int(n) == t.GetTotal() {
+		t.State <- state
+	}
+}
+
+// Get returns the count for a state on a tag
+func (t *Tag) Get(state State) int {
+	var v *uint32
+	switch state {
+	case SPLIT:
+		v = &t.split
+	case STORED:
+		v = &t.stored
+	case SENT:
+		v = &t.sent
+	case SYNCED:
+		v = &t.synced
+	}
+	return int(atomic.LoadUint32(v))
+}
+
+// GetUid returns the unique identifier
+func (t *Tag) GetUid() uint32 {
+	return t.uid
+}
+
+func (t *Tag) GetName() string {
+	return t.name
+}
+
+// GetTotal returns the total count
+func (t *Tag) GetTotal() int {
+	return int(atomic.LoadUint32(&t.total))
+}
+
+// SetTotal sets total count to SPLIT count
+// is meant to be called when splitter finishes for input streams of unknown size
+func (t *Tag) SetTotal() int {
+	total := atomic.LoadUint32(&t.split)
+	atomic.StoreUint32(&t.total, total)
+	return int(total)
+}
+
+// Status returns the value of state and the total count
+func (t *Tag) Status(state State) (int, int) {
+	return t.Get(state), int(atomic.LoadUint32(&t.total))
```

Comment:
this comparison logic should be changed slightly. Probably should also return error (`not available yet`)

For `SPLIT` should be compared to `TOTAL` if that is set, otherwise error

For `STORED` the logic needs to change, `PUT` should increment `STORED` or `SEEN` depending on the return bool. 

`SENT` and `SYNCED` should compare to  `STORED` but only when  `TOTAL` is set and `STORED` +`SEEN` == `TOTAL`.
(assuming `DoneSplit` is called when chunker split ends` or we set the total at creation)

Reply:
... what is `SEEN`?

Reply:
`SEEN` or `FOUND` counts the duplicate chunks, those that are found in db and have been pushed  already.

We could also just increment `STORED`, `SENT` and `SYNCED`  when a chunk is found, but

- that would be wrong semantically (if found then not actually sent)
- would not give accurate ETA
- would need to be done in `hasherstore` or chunker which should not know about push sync   or its related states

Reply:
^ this needs  to be done as part of `Status`, it should also have a third error type return value

Reply:
i pushed the change, but i don't like the ambiguity of the meaning of `total`. in some cases it is `STORED` and in some cases it's the `TOTAL`. maybe we should just return an unsigned int between `0` and `100` or `1000` to signal the progress/status.